### PR TITLE
rgw: simplify and enable by default the transfer accounting.

### PR DESCRIPTION
**Pull request by rzarzynski**
This PR is a very early PoC of RadosGW's accounting simplification. It bases on https://github.com/ceph/ceph/pull/16834 and requires extensive testing. At the moment it was validated only with brief Tempest run.

### CODE REVIEW

**Comment by rzarzynski** [comment]
Jenkins retest this please (`Makefile failed to find target 'tests'`)

**Comment by rzarzynski** [comment]
Jenkins retest this please (`run-cli-tests`)

File: src/rgw/rgw_rest.cc [hunk 1]
```diff
@@ -1139,18 +1138,17 @@ int RGWPutObj_ObjStore::get_data(bufferlist& bl)
 
   int len = 0;
   if (cl) {
-    ACCOUNTING_IO(s)->set_account(true);
+    /* There is no need to manage the accounting enablement here. It's now
```
Comment (amitkumar50) [inline src/rgw/rgw_rest.cc]: As per coding guidelines Please use "It is" instead of it's.
A PR is already raised and approved for it.

Comment (rzarzynski) [inline src/rgw/rgw_rest.cc]: Sure, changed. Thanks for you review!

Comment (cbodley) [inline src/rgw/rgw_rest.cc]: @amitkumar50 to which guidelines or pr are you referring? i only see https://github.com/ceph/ceph/pull/16705, which is only about ceph docs - not code comments

File: src/rgw/rgw_asio_frontend.cc [hunk 2]
```diff
@@ -109,13 +109,13 @@ static void handle_connection(RGWProcessEnv& env, tcp::socket socket,
     RGWRequest req{env.store->get_new_req_id()};
 
     rgw::asio::ClientIO real_client{socket, parser, buffer};
-
+    auto acct_client_io = rgw::io::add_accounting(cct, &real_client);
```
Comment (amitkumar50) [inline src/rgw/rgw_asio_frontend.cc]: @rzarzynski Tab space left before auto.
Though we need spacebars spaces

File: src/rgw/rgw_client_io_filters.h [hunk 3]
```diff
@@ -246,8 +153,8 @@ size_t BufferingFilter<T>::complete_request()
     }
     data.clear();
     buffer_data = false;
-    lsubdout(cct, rgw, 30) << "BufferingFilter::complete_request: buffer_data: sent="
-        << sent << dendl;
+    lsubdout(cct, rgw, 30) << "BufferingFilter::complete_request: buffer_data:"
+                           << " sent=" << sent << dendl;
```
Comment (amitkumar50) [inline src/rgw/rgw_client_io_filters.h]: @rzarzynski tabs left

**Comment by rzarzynski** [comment]
#16834  has brought https://github.com/ceph/ceph/pull/16834/commits/a45ab45f74042617ec382f3d50f4d45d883042f5 which extends the verification of usage accounting at Sepia. It seems that all we need to reliably test changes like this one is just a Teuthology run.

File: src/rgw/rgw_client_io.h [hunk 4]
```diff
@@ -332,22 +332,41 @@ class StaticOutputBufferer : public std::streambuf {
...
   std::vector<std::shared_ptr<DecoratedRestfulClient>> filters;
 
 public:
   ~RGWRestfulIO() override = default;
 
-  RGWRestfulIO(CephContext *_cx, rgw::io::RestfulClient* engine)
-    : AccountingFilter<rgw::io::RestfulClient*>(_cx, std::move(engine)) {
+  RGWRestfulIO(CephContext* const ctx,
+               rgw::io::RestfulClient* engine,
+               rgw::io::Accounter* const accounter)
```
Comment (cbodley) [inline src/rgw/rgw_client_io.h]: seems like this argument should be passed by reference instead, if it's being dereferenced immediately and stored as a reference?

File: src/rgw/rgw_client_io.h [hunk 5]
```diff
@@ -332,22 +332,41 @@ class StaticOutputBufferer : public std::streambuf {
...
+    : DecoratedRestfulClient<rgw::io::RestfulClient*>(std::move(engine)),
```
Comment (cbodley) [inline src/rgw/rgw_client_io.h]: i know this part isn't new, but why `std::move()` when `engine` is a pointer?

Comment (rzarzynski) [inline src/rgw/rgw_client_io.h]: It looks strange but it necessary as long as the parent class' constructor requires r-value reference. It does that to explicitly inform the next stage never should be accessed directly but with assistance of `DecoratedRestfulClient` instead. I will add a comment.

File: src/rgw/rgw_client_io.h (hunk 4 above)
Comment (rzarzynski) [inline src/rgw/rgw_client_io.h]: It is. Switching to reference.

**Comment by rzarzynski** [comment]
Rebased after a trivial merge conflict.

File: src/rgw/rgw_client_io.h (hunk 5 above)
Comment (cbodley) [inline src/rgw/rgw_client_io.h]: i see, because the template argument is a pointer - thanks

**Comment by mattbenjamin** [comment]
@rzarzynski can you find a moment to rebase?
@cbodley re-review?

**Pull request closed without merging**